### PR TITLE
Issue #7764: Update AbstractChecks to log DetailAST - UncommentedMain

### PR DESCRIPTION
**Pull request by AmrDeveloper**
Issue #7764 

Update AbstractChecks to log DetailAST - UncommentedMain

Report : https://amrdeveloper.github.io/checkstyle-reports/7764/

### CODE REVIEW

File: src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionUncommentedMainTest.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    private final Class<UncommentedMainCheck> clazz =
+            UncommentedMainCheck.class;
+
+    @Override
+    protected String getCheckName() {
+        return clazz.getSimpleName();
+    }
+
+    @Test
+    public void testOne() throws Exception {
```
Comment (esilkensen) [inline src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionUncommentedMainTest.java]: maybe this test could have a more descriptive name?

Comment (romani) [reply src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionUncommentedMainTest.java]: we did not find any good names for them, that why it is like this in all Xpath tests.

@esilkensen , do you have suggestions ?

Comment (romani) [inline src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionUncommentedMainTest.java]: @AmrDeveloper , please add one more test, we need at least two tests.

**Comment by romani** [comment]
> I'm not sure what in this PR triggered the pitest failure in SuppressWarningsHolder

strange .... I relaunched it, if is appears again and become persistent, we need to find a workaround. We cannot merge PR that cause CI failure.

File: src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionUncommentedMainTest.java (hunk 1 above)
Comment (esilkensen) [reply src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionUncommentedMainTest.java]: Ah, this sounds good to me then - I didn’t realize it was the same in all other tests too.

Comment (AmrDeveloper) [reply src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionUncommentedMainTest.java]: @romani Done but CI not passed again !

Comment (esilkensen) [reply src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionUncommentedMainTest.java]: I don’t think the pitest failure is transient/flaky, but I tried looking at how to resolve it yesterday and didn’t find the answer...

A common approach could be to take the line of code and the mutation that is surviving and construct a test that fails when that mutation is present, but the confusing thing here is pitest lists two identical mutations for the same line with one surviving, so I’m not sure what the report is telling us.

Comment (romani) [reply src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionUncommentedMainTest.java]: > but CI not passed again !

it is even better, at least behavior is stable. It is sad that it blocks your PR. Looks like we run into pitest bug, so we need to find a workaround....

Comment (romani) [reply src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionUncommentedMainTest.java]: it does not mean it is good in all others :) , we just do not know better alternatives.

reason of this tests is not to validate Xpath correctness and make sure it works in general, that is why names are so general. If in future we will run into problems of xpath for particular Check then we will create test like `testOnInnerClasswithAnnotationAndXxxxxx`.

**Comment by romani** [comment]
@AmrDeveloper , please rebase one more time on our latest code, I just merged fix to publish pitest report in CircleCI, it would help to investigate problem by mentors (as we become able to see report from mobile phones and think about this issue on the go)

**Comment by romani** [comment]
@AmrDeveloper , after rebase, please take some other issue meanwhile, this PR will hang for a bit to let us think on what to do.

**Comment by rnveach** [comment]
https://40666-12499251-gh.circle-artifacts.com/2/target/pit-reports/202003142034/com.puppycrawl.tools.checkstyle.checks/SuppressWarningsHolder.java.html#org.pitest.mutationtest.report.html.SourceFile@46855114_246
> || entry.getFirstLine() == line
> removed conditional - replaced equality check with true → SURVIVED

I replaced the code locally and this is the failure I got:
`SuppressWarningsFilterTest.testDefault:125->verifySuppressed:131->verifySuppressed:175->AbstractModuleTestSupport.verify:220->AbstractModuleTestSupport.verify:256->AbstractModuleTestSupport.verify:292 error message 0 expected:<[M:\checkstyleWorkspace\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\filters\suppresswarningsfilter\InputSuppressWarningsFilter.java:16: Missing a Javadoc comment].> but was:<[Audit done].>`

**Comment by rnveach** [comment]
@romani I am unable to reproduce the issue.

One way to fix this quickly is just to add reflection tests to `SuppressWarningsHolder.isSuppressedAfterEventStart` to also kill the mutation.

**Comment by romani** [comment]
@AmrDeveloper ,

if we have some signal from pitest that mutation is not covered .... and message is not clear ... we need to refactor code.
We have now:
```
return entry.getFirstLine() < line
  || entry.getFirstLine() == line
  && (column == 0 || entry.getFirstColumn() <= column);
```
please move all expressions to local variables, like:
```
 return var1 && var2 || (var3 || var4)
```
and keep extracting till pitest be able to give us clear message.
Please let me know is this let find root reason. You can share reports at github.io - https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#deploying-report .

**Comment by AmrDeveloper** [comment]
@romani Thanks for your suggestion it helped me to find the reason

the problem is in 

    column == 0

we don't need to check if column is equal  0 we can just put and it will check that it will be after column in the same line

    entry.getFirstLine() == line && entry.getFirstColumn() <= column

The full code will be 

    return entry.getFirstLine() < line
                 || entry.getFirstLine() == line
                 && entry.getFirstColumn() <= column;

It passed all in SuppressWarningsHolderTest and passed ./.ci/pitest.sh pitest-misc without any different output 

passed

Should i fix it in current PR or create another one and after merge it i can rebase current PR ?

**Comment by esilkensen** [comment]
Nice find @AmrDeveloper!

Since the pitest failure only seems to have surfaced as a result of the other changes in this PR, it would make sense to me for your fix/refactoring to go in this PR - but I defer to @romani @rnveach and others for the decision.

**Comment by rnveach** [comment]
> we don't need to check if column is equal 0
>  && entry.getFirstColumn() <= column;
> It passed all in SuppressWarningsHolderTest and passed ./.ci/pitest.sh pitest-misc without any different output

My local on a clean master fails with this change.
````
[INFO] Running com.puppycrawl.tools.checkstyle.filters.SuppressWarningsFilterTest
[ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.045 s <<< FAILURE! - in com.puppycrawl.tools.checkstyle.filters.SuppressWarningsFilterTest
[ERROR] testDefault  Time elapsed: 0.022 s  <<< FAILURE!
org.junit.ComparisonFailure: error message 9 expected:<...arningsFilter.java:7[6: Missing a Javadoc comment].> but was:<...arningsFilter.java:7[1: Uncommented main method found].>
````

Did you run it on a clean master?

It seems more likely to me this whole issue is coming to light as we convert more checks to log AST in this `SuppressWarningsHolderTest`. It was an on and off problem before and now it is completely on in this PR just AS we are converting another check to log AST. Everything in the test is an AST log except what was once this check and 1 other. It seems to me this test doesn't test enough cases anymore for non-AST loggers.

**Comment by esilkensen** [comment]
Oh, yes, sorry I spoke before trying the code myself...

Not related to PR, but looking at it now I am confused by the name/javadoc which says that `isSuppressedAfterEventStart` returns `true if suppression entry position is after the audit event occurrence position in the source file`  but

```java
        return entry.getFirstLine() < line
            || entry.getFirstLine() == line
            && (column == 0 || entry.getFirstColumn() <= column);
```

looks like it checks whether the audit event occurrence (`line` and `column`) comes after the suppression `entry` (except a special case if they are on the same line and the event has column 0)?

**Comment by esilkensen** [comment]
I just tried cloning @AmrDeveloper's branch and making only this change
```java
private static boolean isSuppressedAfterEventStart(int line, int column, Entry entry) {
    final boolean var1 = entry.getFirstLine() < line;
    final boolean var2 = entry.getFirstLine() == line;
    final boolean var3 = column == 0;
    final boolean var4 = entry.getFirstColumn() <= column;
    return var1 || var2 && (var3 || var4);
}
```
and as a result `./.ci/pitest.sh pitest-misc` reported no errors

**Comment by rnveach** [comment]
> except a special case if they are on the same line and the event has column 0

It has to do with logging a violation on a line and a line and column. If there is no column number, it is probably defaulted to 0 instead of null. I don't know where the the values are coming from, but if the main pattern doesn't have a column position it just wants to match the violation to just the line only.

**Comment by esilkensen** [comment]
Ah, that makes sense.

Anyway, as far as I can tell, CI may be green by taking this PR as it is now and simply storing `entry.getFirstLine() == line` in a variable, like
```java
private static boolean isSuppressedAfterEventStart(int line, int column, Entry entry) {
    final boolean onSameLine = entry.getFirstLine() == line;
    return entry.getFirstLine() < line
        || onSameLine && (column == 0 || entry.getFirstColumn() <= column);
}
```

**Comment by AmrDeveloper** [comment]
@rnveach yes you are right it fails on master because UncommentedMainCheck logging just have line number and column number will be 0 as default value and it passed on my branch because logging now have line and column number

**Comment by romani** [comment]
@AmrDeveloper , thanks a lot for investigation.

I crested https://github.com/checkstyle/checkstyle/issues/7885
Lets fix it first,  in separate PR, and then rebase.

**Comment by esilkensen** [comment]
I think CI for this PR can be green by just extracting `entry.getFirstLine() == line` into a variable (following your debugging suggestion, @romani) - no need to change the logic (see my last comment https://github.com/checkstyle/checkstyle/pull/7836#issuecomment-599807606) - I think this looks like a quirk/bug of pitest.

I think what @rnveach pointed out is that if you do change the logic by removing the check for `column == 0`, then a test fails in SuppressWarningsFilterTest (note this is different than SuppressWarningsHolderTest)

**Comment by romani** [comment]
> by just extracting entry.getFirstLine() == line into a variable 

I am ok with that. 
We need special issue for this, to keep reports of reason why this code exists.
@strkkk likes to inline all in expressions, he might be interested to fix it back.

**Comment by AmrDeveloper** [comment]
@romani @esilkensen Thanks i will fix #7885 first using this way #7836 (comment)

**Comment by esilkensen** [comment]
Oh, wait I'm sorry, I might have confused myself and been running pitest from master instead of your branch... so, maybe the variable renaming doesn't resolve the issue after all. I will try this out quickly to confirm...

**Comment by esilkensen** [comment]
@AmrDeveloper sorry for the confusion, now I think (and have confirmed on your branch 🙂) that pitest passes after adding an additional test case for SuppressWarningsHolder that exercises that `column == 0` path, no changes needed to the variable names or logic. For example:

```java
@Test
public void testIsSuppressedAfterEventStart2() throws Exception {
    populateHolder("check", 100, 100, 350, 350);
    final AuditEvent event = createAuditEvent("check", 100, 0);

    assertTrue(SuppressWarningsHolder.isSuppressed(event), "Event is not suppressed");
}
```

**Comment by AmrDeveloper** [comment]
@esilkensen Nice suggestion thanks 🙂

**Comment by romani** [comment]
@esilkensen ,

> testIsSuppressedAfterEventStart2

we discourage creation of fake tests, it is allowed only as absolutely last resort, and there should be big comment above method why should be it in fake way. 
All tests should be based in Input files and execution of whole checkstyle.

**Comment by rnveach** [comment]
@AmrDeveloper Other PR was merged, please rebase this so we can get a green CI.

**Comment by AmrDeveloper** [comment]
@rnveach Thanks and done

**Pull request merged**